### ISSUE TITLE
Question about the adjacent matrix

### ISSUE THREAD

**Issue by speedcell4** [Cartus/AGGCN_TACRED#2]
Hi~

In your abstract, you said "In this work, we propose Attention Guided Graph Convolutional Networks (AGGCNs), a novel model which directly takes full dependency trees as inputs", but I don's see how do you make use of the dependency tree.

According to Equation (2), seems you just construct the $\tilde{A}$ by using the input embeddings, so where do you make use of the original adjacency matrix $A$?

In your code, I still cannot confirm how do you make use of $A$, which is `adj` in your code, right?
https://github.com/Cartus/AGGCN_TACRED/blob/master/model/aggcn.py#L175

Could you please show me the details?
And by the way, that is the matrix $V$ in your Equation (2)? I can not find the definition of $V$ neither.

**Comment by speedcell4** [comment]
Yes, I know you made `mask` by using `adj`, but isn't this equivalent to `src_mask` above?

**Comment by Cartus** [comment]
Hi,

For the first question, there is a description in the paper:

>  In practice, we treat the original adjacency matrix as an initialization so that the dependency information can be captured in the node representations for later attention calculation. The attention guided layer is included starting from the second block.

**Comment by Cartus** [comment]
I close this issue if you have any further questions feel free to reopen it.

**Comment by speedcell4** [comment]
But why do you need to double your sub-layers? I mean your `sublayer_first ` and `sublayer_second`?

**Comment by speedcell4** [comment]
And there is not **reopen** button in this page

**Comment by Cartus** [comment]
For the sub-layer problem, u can refer to this [TACL paper-DCGCN](http://www.statnlp.org/paper/2019/densely-connected-graph-convolutional-networks-graph-sequence-learning.html). Basically, the motivation behind this is to imitate convolution filters of different sizes (1x1, 3x3, etc.) in CNN.

The number of sub-layers in each block is different for TACRED. Here the first sub-layer is 2, the second is 4. You can refer to the train.py. These are hyper-parameters.

**Comment by speedcell4** [comment]
Thank you~

**Comment by ardellelee** [comment]
Hello,  I still have some questions about the number of layers. 

In your AGGCN paper, Section 3.2, you mentioned that the best setting for sentence-level relation extraction is M=2, L=5.  
- Is L the sum of sublayer_first and sublayer_second?
- As I understand, M is the number of AGGCN blocks, and should be identical to the argument  --num_layers in train.py. However, the argument is indicated as "Num of RNN layers" in the code. A bit confused. Is it a typo?

> For the sub-layer problem, u can refer to this [TACL paper-DCGCN](http://www.statnlp.org/paper/2019/densely-connected-graph-convolutional-networks-graph-sequence-learning.html). Basically, the motivation behind this is to imitate convolution filters of different sizes (1x1, 3x3, etc.) in CNN.
> 
> The number of sub-layers in each block is different for TACRED. Here the first sub-layer is 2, the second is 4. You can refer to the train.py. These are hyper-parameters.

**Comment by Cartus** [comment]
Hi ,
> * Is L the sum of sublayer_first and sublayer_second?

Yes, you are correct.

> * As I understand, M is the number of AGGCN blocks, and should be identical to the argument  --num_layers in train.py. However, the argument is indicated as "Num of RNN layers" in the code. A bit confused. Is it a typo?

Yes, it is a typo... Thank you so much for pointing it out! It should be the number of blocks which is M.

**Comment by TranVanHien** [comment]
Hello @Cartus , you mentioned that the best setting for sentence-level relation extraction is M=2, L=5. L the sum of sublayer_first and sublayer_second, with the first sub-layer is 2, the second is 4. But 2+4 = 5 ?

**Comment by TranVanHien** [comment]
Hi @speedcell4 , could you explain for me? :)

**Comment by speedcell4** [comment]
@TranVanHien Sorry, I can not, I realized I can never figure out these complicated experiment settings and then I decided to give up doing relation extraction task two months ago

**Comment by TranVanHien** [comment]
Thank you.

**Comment by Cartus** [comment]
Hi @TranVanHien , that is a typo in the paper, I will fix it later. The code in this repo is the default setting for the TACRED dataset. 

Thank you for pointing it out!

**Comment by TranVanHien** [comment]
I see. Thank you for your quick reply.

**Comment by liuyijiang1994** [comment]
Hi, Cartus @Cartus 
I read your answers about the use of dependency tree above, you takes it as an initialization in your work, so it just works in the first GCN layer, right? If so, thence the second GCN layer, the adj matrix becomes the weight produced by self attention, and structure becomes to a full strongly connected directed graph, is there any difference between the calculation of self attention and the attention GCN in a full strongly connected directed graph?
Thanks for your help!

**Comment by Cartus** [comment]
Hi @liuyijiang1994 ,

Thanks for the question. Actually they are basically the same. I read a paper few weeks ago. The paper discusses about the GNNs and Transformer model, which might be useful. They also have a nice summary of these works including different GNNs and different self-attention methods.

Contextualized Non-local Neural Networks for Sequence Learning
https://arxiv.org/abs/1811.08600#

**Comment by liuyijiang1994** [comment]
@Cartus  Thank you for your prompt reply, it is very helpful to me!
